### ISSUE TITLE
WAR support

### ISSUE THREAD

**Issue by yrashk** [lburgazzoli/gradle-karaf-plugin#13]
In my OSGi application, I have mostly regular bundles (JARs) and just one war. Right now, the plugin happily puts the JAR into the KAR, however, this is not what I want it to do. I either have to leave the dependency out of the feature file and then install it manually, but I'd rather automate this.

I think the issue is twofold:

1. Feature needs to be generated properly, here's an example of using a war file:

```xml
 <feature name="activemq-web-console" version="5.4.2">
        <bundle>war:mvn:org.apache.activemq/activemq-web-console/5.4.2/war?Webapp-Context=activemq-web-console</bundle>
    </feature>
```

(Found it by googling at http://karaf.922171.n3.nabble.com/Installing-a-war-file-in-a-feature-td2747919.html)

2. At least in my case, building a proper war requires invoking the "war" task (instead of "jar"). I am still learning this plugin's implementation to see how feasible this would be.

Thoughts?

**Comment by lburgazzoli** [comment]
In the old version it was supported, let me check

**Comment by yrashk** [comment]
It is actually more complex, I need this particular component to be
included BOTH as a regular bundle (jar, for interfaces to be used by other
components, primarily) and a war (war). They will have different symbol
names.

On Mon, Mar 21, 2016 at 7:47 AM, Luca Burgazzoli <notifications@github.com>
wrote:

> In the old version it was supported, let me check
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lburgazzoli/gradle-karaf-plugin/issues/13#issuecomment-199322214>
>



-- 
Y.

**Comment by lburgazzoli** [comment]
are you in multi-project setup ? i.e. the feaure plugin is applied only to a subproject

**Comment by yrashk** [comment]
Yes, it's a multi-project setup.

On Mon, Mar 21, 2016 at 8:32 AM, Luca Burgazzoli <notifications@github.com>
wrote:

> are you in multi-project setup ? i.e. the feaure plugin is applied only to
> a subproject
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lburgazzoli/gradle-karaf-plugin/issues/13#issuecomment-199342130>
>



-- 
Y.

**Comment by yrashk** [comment]
Any idea at least where I can start with this?

On Mon, Mar 21, 2016 at 9:03 AM, Yurii Rashkovskii <yrashk@gmail.com> wrote:

> Yes, it's a multi-project setup.
>
> On Mon, Mar 21, 2016 at 8:32 AM, Luca Burgazzoli <notifications@github.com
> > wrote:
>
>> are you in multi-project setup ? i.e. the feaure plugin is applied only
>> to a subproject
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/lburgazzoli/gradle-karaf-plugin/issues/13#issuecomment-199342130>
>>
>
>
>
> --
> Y.
>
>


-- 
Y.

**Comment by lburgazzoli** [comment]
I've added better support for WAR and multi project setups in 0.0.18 however I do not know how to handle the jar + war scenario, for the plugin to work it would be enough to have the project listed twiice with different scope but I do not know how to do so in gradle.

**Comment by yrashk** [comment]
How do I do the war (no jar) scenario?

On Mon, Mar 21, 2016 at 9:51 AM, Luca Burgazzoli <notifications@github.com>
wrote:

> I've added better support for WAR and multi project setups in 0.0.18
> however I do not know how to handle the jar + war scenario, for the plugin
> to work it would be enough to have the project listed twiice with different
> scope but I do not know how to do so in gradle.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lburgazzoli/gradle-karaf-plugin/issues/13#issuecomment-199374597>
>



-- 
Y.

**Comment by lburgazzoli** [comment]
Well, one thing has improved one not yet :-)

If you have a WAR in your dependencies included as:
 ```
    compile("org.apache.activemq:activemq-web-console:5.13.2@war") 
 ```

Then the plugin will render the url properly however it can't handle war dependencies as projects at the moment

**Comment by yrashk** [comment]
I just decided to switch to a plain JAR, I think that works even better for
me (WARs lose Service-Components for some reason with Gradle war plugin)

On Mon, Mar 21, 2016 at 10:32 AM, Luca Burgazzoli <notifications@github.com>
wrote:

> Well, one thing has improved one not yet :-)
>
> If you have a WAR in your dependencies included as:
>
>     compile("org.apache.activemq:activemq-web-console:5.13.2@war")
>
> Then the plugin will render the url properly however it can't handle war
> dependencies as projects at the moment
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lburgazzoli/gradle-karaf-plugin/issues/13#issuecomment-199391388>
>



-- 
Y.

**Comment by lburgazzoli** [comment]
I will keep this ticket open to check how to improve war deps management

**Comment by lburgazzoli** [comment]
Solved in v0.0.23

```
$ jar tvf features/build/karaf/kar/features-1.0.0.kar | egrep '.xml|.jar|.war'
2308517 Fri Dec 11 11:34:46 CET 2015 repository/com/google/guava/guava/19.0/guava-19.0.jar
   767 Wed Mar 30 14:52:24 CEST 2016 repository/com/github/lburgazzoli/sandbox/gradle/karaf/features/1.0.0/features-1.0.0.xml
   968 Wed Mar 30 14:52:22 CEST 2016 repository/com/github/lburgazzoli/sandbox/gradle/karaf/features/1.0.0/features-1.0.0.jar
   988 Wed Mar 30 14:52:24 CEST 2016 repository/com/github/lburgazzoli/sandbox/gradle/karaf/module1/1.0.0/module1-1.0.0.jar
468200 Wed Mar 30 14:52:24 CEST 2016 repository/com/github/lburgazzoli/sandbox/gradle/karaf/module2/1.0.0/module2-1.0.0.war
   938 Wed Mar 30 14:52:24 CEST 2016 repository/com/github/lburgazzoli/sandbox/gradle/karaf/module2/1.0.0/module2-1.0.0.jar
 33088 Wed Dec 09 12:12:50 CET 2015 repository/org/slf4j/slf4j-api/1.7.13/slf4j-api-1.7.13.jar
 42375 Mon Dec 07 14:36:36 CET 2015 repository/org/slf4j/slf4j-ext/1.7.13/slf4j-ext-1.7.13.jar
434678 Thu Nov 19 10:59:10 CET 2015 repository/org/apache/commons/commons-lang3/3.4/commons-lang3-3.4.jar
```
